basic_module/distribution/src/impls.rs

```diff
@@ -0,0 +1,114 @@
+        block_author: Public,
+        block_hash: BlockHash,
+        rewards: BTreeMap<Public, u64>,
+    ) -> Result<(), Error> {
+        self.block_number = block_number;
+        self.block_author = block_author;
+        let mut intermediate_rewards = get_intermediate_rewards(get_key());
+
+        let term = current_term_id(block_number);
+
+        if block_number == last_term_finished_block_num(block_number) + 1 {
+            intermediate_rewards.update_current_rewards_to_empty()?;
+            match term {
+                0 | 1 => {}
+                _ => {
+                    let banned = get_banned_validators();
+                    let start_of_the_current_term_header =
+                        get_block_header(&block_hash).ok_or(Error::InvalidHash(block_hash))?;
+
+                    let pending_rewards =
+                        calculate_pending_rewards_of_the_term(banned, rewards, start_of_the_current_term_header)?;
+                    intermediate_rewards.calculated_reward = pending_rewards.into_iter().collect();
+                    intermediate_rewards.save_to_state();
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl CloseBlockHandler for Distribution {
+    fn distribute(&self, total_reward: u64, total_min_fee: u64, stakes: HashMap<Public, u64>) -> Result<(), Error> {
+        let mut intermediate_rewards = get_intermediate_rewards(get_key());
+
+        let block_author = self.block_author;
+        let block_number = self.block_number;
+
+        assert!(total_reward >= total_min_fee, "{} >= {}", total_reward, total_min_fee);
+
+        let mut distributor = fee_distribute(total_min_fee, &stakes);
+        for (address, share) in &mut distributor {
+            add_balance(&address, share);
+        }
+
+        let block_author_reward = total_reward - total_min_fee + distributor.remaining_fee();
+        let term = current_term_id(block_number);
+        match term {
+            0 => {
+                add_balance(&block_author, block_author_reward);
+            }
+            _ => {
+                intermediate_rewards.update_quantity(&block_author, block_author_reward);
+                intermediate_rewards.save_to_state();
+            }
+        }
+
+        if is_term_changed(self.block_number) {
+            for (address, reward) in intermediate_rewards.drain_calculated_rewards() {
+                add_balance(&address, reward);
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl FeeManager for Distribution {
+    fn accumulate_block_fee(&self, total_additional_fee: u64, total_min_fee: u64) {
+        let context = get_context();
+        let mut intermediate_rewards: IntermediateRewards = get_intermediate_rewards(get_key());
+
+        intermediate_rewards.total_additional_fee += total_additional_fee;
+        intermediate_rewards.total_min_fee += total_min_fee;
+
+        context.set(get_storage_id(), &get_key(), intermediate_rewards.into());
```

Comment:
Why don't you use `intermediate_rewards.save_to_state()`?
One of the two code snippets should be changed.